File: b01260.py
from collections import deque

def dfs(graph, v, visited):
    visited[v] = True
    print(v, end=" ")
    
    for neighbor in sorted(graph[v]):  # child node 중 작은 번호부터 방문
        if not visited[neighbor]:
            dfs(graph, neighbor, visited)

def bfs(graph, start):
    visited = {node: False for node in graph}
    queue = deque([start]) # 방문할 리스트
    visited[start] = True
    
    while queue:
        v = queue.popleft()
        print(v, end=" ")
        
        for neighbor in sorted(graph[v]):  # 작은 번호부터 방문
            if not visited[neighbor]:
                queue.append(neighbor) # 방문할 리스트에 모든 자식 노드들 추가
                visited[neighbor] = True

# 입력 받기
n, m, v = map(int, input().split()) # node, edge root. e.g 4, 5, 1
# map : iterator. 한 번에 하나의 요소만 반환. 필요할 때만 값 생성 (lazy evaluation)
graph = {i: set() for i in range(1, n + 1)}

for _ in range(m):
    a, b = map(int, input().split())
    graph[a].add(b)
    graph[b].add(a)  # 양방향 연결

# DFS 실행
visited_dfs = {node: False for node in graph}
dfs(graph, v, visited_dfs)
print()

# BFS 실행
bfs(graph, v)

File: b01260.cpp
#include <iostream>
#include <queue>

int am[1001][1001];
int visited[1001];

int n; int m; int v;

void dfs(int v){
    visited[v] = 1;
    std::cout << v << " "; 
    for (int i=1; i<=n; i++){
        if (am[v][i] == 1 && visited[i] == 0) dfs(i);
    }
}

std::queue<int> q;

void visit(int v){
    q.push(v);
    visited[v]=1;
    std::cout << v << " ";
}

void bfs(int v){
    visit(v);
    while (!q.empty()){
        int w = q.front();
        q.pop();
        for (int i=1; i<=n; i++){
            if (am[w][i] == 1 && visited[i] == 0) visit(i);
        }
    }
}

int main(){
    std::cin >> n;
    std::cin >> m;
    std::cin >> v;
    
    for (int i=1; i<1001; i++){
        visited[i]=0;
        for (int j=1; j<1001; j++){
            am[i][j]=0;
        }
    }
    for (int i=1; i<=m; i++){
        int j; int k;
        std::cin >> j;
        std::cin >> k;
        am[j][k] = 1;
        am[k][j] = 1;
    }

    dfs(v);
    for (int i=1; i<1001; i++) visited[i]=0;
    std::cout << std::endl;
    bfs(v);

    return 0;
}